### PR TITLE
Drivetrain clean

### PR DESCRIPTION
**Pull request by wollip**
Integrated changes from hardware overhaul

### CODE REVIEW

File: include/configs.h [hunk 1]
```diff
@@ -33,10 +33,9 @@
 #define ZIPLINE_TURN_DEGREES() 75
 
 // PID initial values
-#define GAIN_T() 1
-#define GAIN_P() 12
-#define GAIN_I() 0
-#define GAIN_D() 0
+#define GAIN_P() 12.0
```
Comment (wollip) [inline include/configs.h:36]: what are these use for?

File: include/configs.h [hunk 2]
```diff
@@ -110,6 +110,21 @@
...
 #define PICKUP_FORWARD_DISTANCE() 10
 
+#define LINE_FOLLOW_0_SPEED() 120
+#define LINE_FOLLOW_0_GAIN() 24
+#define LINE_FOLLOW_1_SPEED() 255
+#define LINE_FOLLOW_1_GAIN() 50
+#define LINE_FOLLOW_2_SPEED() 255
+#define LINE_FOLLOW_2_GAIN() 50
+
+/*#define LINE_FOLLOW_0_SPEED() 80
```
Comment (wollip) [inline include/configs.h]: leave a message on why these are here

File: include/hardware/encoder.h [hunk 3]
```diff
@@ -1,43 +1,30 @@
-#ifndef WALKERVILLE_ROBOT_HARDWARE_ENCODER_H
-#define WALKERVILLE_ROBOT_HARDWARE_ENCODER_H
-
+#pragma once
 #include "configs.h"
-
 #include <stdint.h>
+#include "Print.h"
```
Comment (wollip) [inline include/hardware/encoder.h]: warp around with debug

File: include/hardware/encoder.h [hunk 4]
```diff
@@ -1,43 +1,30 @@
...
-   * @param distance The distance in centimeters.
-   * @return The equivalent number of encoder ticks.
-   */
-  static uint32_t cmToTicks(unsigned int distance);
+  Encoder(uint8_t encoderId, bool reverse = false);
+  inline ~Encoder() {};
+  int32_t getPosition() const;
+  int16_t getVelocity() const;
+  void tick();
+  void printTo(Print& p);
```
Comment (wollip) [inline include/hardware/encoder.h]: debug

File: include/hardware/encodermotor.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+private:
+  bool m_backward;
+  int m_motor_num;
+  Encoder& m_encoder;
+
+  const static int32_t deadZone = 5;
+
+  enum ControlMode { IDLE, POWER, VELOCITY, POSITION };
+  ControlMode m_controlMode;
+  static const int16_t m_accelerationLimit = 20;
```
Comment (wollip) [inline include/hardware/encodermotor.h]: move to CPP

File: include/hardware/encodermotor.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  void hold();
+  void tick();
+  int32_t getPosition();
+
+private:
+  bool m_backward;
+  int m_motor_num;
+  Encoder& m_encoder;
+
+  const static int32_t deadZone = 5;
```
Comment (wollip) [inline include/hardware/encodermotor.h]: move to CPP

File: include/sequences/drivetrain.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  void commandTurnRight(float angle, int16_t speed = 20);
+  //void commandTurnPivot(int32_t distance, int16_t speed = 255);
+  void commandLineFollow(uint8_t speedSetting);
+  void commandBeaconFollow();
+  void stop();
+  bool readyForCommand();
+  void setPower(int16_t left, int16_t right);
+  void setVelocity(int16_t left, int16_t right);
+  void tick();
+  void printTo(Print& p);
```
Comment (wollip) [inline include/sequences/drivetrain.h]: debug

File: include/utils/pid.h [hunk 8]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+#include <Print.h>
```
Comment (wollip) [inline include/utils/pid.h]: debug

File: include/utils/pid.h [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+class PidController {
+public:
+  PidController(float kP, float kI, float kD, float kDecay = 0.9);
+  inline ~PidController() {};
+  float kP;
+  float kI;
+  float kD;
+  float kDecay;
+  float run(float error);
+  void printTo(Print& p);
```
Comment (wollip) [inline include/utils/pid.h]: debug

File: src/hardware/encoder.cpp [hunk 10]
```diff
@@ -1,86 +1,63 @@
...
-
-  volatile uint32_t counts[4];
-  volatile unsigned long prevTime[4];
-  const unsigned long WAITTIME = ENCODER_WAIT_TIME();
-
-Encoder::Encoder(){
-  start(R_ENCODER_);
-  start(L_ENCODER_);
+int32_t Encoder::cmToTicks(float cm) {
+  return cm * GEAR_RATIO() * ENCODER_TICKS_PER_REVOLUTION() / (WHEEL_DIAMETER() * M_PI);
```
Comment (wollip) [inline src/hardware/encoder.cpp]: lets have this as a constant, so we don't have to do the calculation again and again.

File: src/hardware/encoder.cpp [hunk 11]
```diff
@@ -1,86 +1,63 @@
...
-  /* Allow pin to trigger interrupts        */
-  EIMSK |= (1 << INTX);
-  /* Clear the interrupt configuration bits */
-  EICRA &= ~(1 << (INTX*2+0));
-  EICRA &= ~(1 << (INTX*2+1));
-  /* Set new interrupt configuration bits   */
-  EICRA |= mode << (INTX*2);
-  sei();
+int32_t Encoder::degToTicks(float deg) {
+  return cmToTicks((deg / 180.) * AXLE_LENGTH() * M_PI);
```
Comment (wollip) [inline src/hardware/encoder.cpp]: change to constant

File: src/hardware/encoder.cpp [hunk 12]
```diff
@@ -1,86 +1,63 @@
...
+void Encoder::tick() {
+  m_previousPosition = m_currentPosition;
+  m_currentPosition = getPositionPrivate();
+  m_velocitySamples[m_currentVelocitySample++] = m_currentPosition - m_previousPosition;
+  if (m_currentVelocitySample >= m_numVelocitySamples) m_currentVelocitySample = 0;
 }
 
-uint32_t Encoder::cmToTicks(unsigned int distance) {
-  return distance * GEAR_RATIO() * 24 / (WHEEL_DIAMETER() * PI);
+void Encoder::printTo(Print& p) {
```
Comment (wollip) [inline src/hardware/encoder.cpp]: debug

File: src/hardware/encoder.cpp [hunk 13]
```diff
@@ -1,86 +1,63 @@
...
 
-}  // namespace hardware
-
-ISR(INT0_vect){
-  if (millis() > hardware::prevTime[0] + hardware::WAITTIME)
-  {
-    hardware::counts[0]++;
-    hardware::prevTime[0] = millis();
+int32_t Encoder::getPositionPrivate() {
+  if (m_reverse) {
```
Comment (wollip) [inline src/hardware/encoder.cpp]: lets do this in 2 line?

File: src/hardware/encoder.cpp [hunk 14]
```diff
@@ -1,86 +1,63 @@
...
-};
 
-ISR(INT3_vect){
-  if (millis() > hardware::prevTime[3] + hardware::WAITTIME)
-  {
-    hardware::counts[3]++;
-    hardware::prevTime[3] = millis();
-  }
-};
+}
```
Comment (wollip) [inline src/hardware/encoder.cpp]: comment for namespace

File: src/hardware/encodermotor.cpp [hunk 15]
```diff
@@ -0,0 +1,85 @@
+#include "hardware/encodermotor.h"
+#include "phys253.h"
+#include "configs.h"
+
+using namespace utils;
```
Comment (wollip) [inline src/hardware/encodermotor.cpp:5]: i don't like this

File: src/hardware/encodermotor.cpp [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+      int16_t velocity = m_maxVelocity;
+      if (abs(error) < 100) velocity = 5;
+      if (abs(error) < 5) velocity = 0;
+      if (error < 0) m_velocityTarget = velocity;
+      if (error > 0) m_velocityTarget = -velocity;
+      //fall through
+    }
+    case VELOCITY:
+      m_powerTarget += m_velocityPid.run(m_encoder.getVelocity() - m_velocityTarget);
+      if (m_powerTarget > m_maxPower) m_powerTarget = m_maxPower;
```
Comment (wollip) [inline src/hardware/encodermotor.cpp]: lets us `constrain` instead.

File: src/hardware/encodermotor.cpp [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+void EncoderMotor::stop() {
+  m_controlMode = IDLE;
+  motor.stop(m_motor_num);
+}
+
+void EncoderMotor::hold() {
+  setPosition(m_encoder.getPosition(), 20);
+}
+
+#define median(a,b,c) max(min(a,b), min(max(a,b),c))
```
Comment (wollip) [inline src/hardware/encodermotor.cpp]: take this out and replace with `constrain`

File: src/stages/gate.cpp [hunk 18]
```diff
@@ -43,14 +43,14 @@ bool Gate::loop(){
       if (left + right < threshold_) {
 #if CAUTIOUS_GATE_ROUTINE()
         if (!gate_high_){
-          follower_.stop();
+          driver_.stop();
           return false;
         }
 #endif  // CAUTIOUS_GATE_ROUTINE()
-        follower_.loop();
+        driver_.commandLineFollow(true);
```
Comment (lsav) [inline src/stages/gate.cpp]: commandLineFollow should not take a boolean?

File: src/hardware/encodermotor.cpp [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+#include "phys253.h"
+#include "configs.h"
+
+using namespace utils;
+
+namespace hardware {
+
+EncoderMotor::EncoderMotor(int motor_num, Encoder& encoder, bool backward) :
+  m_motor_num(motor_num), m_encoder(encoder), m_backward(backward), 
+  m_velocityPid(-0.4,0.0,0.0), m_positionPid(-0.2,0.0,-0.2)
```
Comment (wollip) [inline src/hardware/encodermotor.cpp]: lets make these configurable?

File: src/hardware/encodermotor.cpp [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+  else return false;
+}
+
+void EncoderMotor::stop() {
+  m_controlMode = IDLE;
+  motor.stop(m_motor_num);
+}
+
+void EncoderMotor::hold() {
+  setPosition(m_encoder.getPosition(), 20);
```
Comment (wollip) [inline src/hardware/encodermotor.cpp]: i believe 20 is the default? lets take out the second arguement

File: src/hardware/qrd.cpp [hunk 21]
```diff
@@ -21,14 +21,6 @@ int8_t Qrd::getTapeError(){
   const int rc_val = analogRead(RC_QRD_SENSOR_);
   const int rr_val = analogRead(RR_QRD_SENSOR_);
 
-#if DEBUG()
```
Comment (wollip) [inline src/hardware/qrd.cpp]: should we keep this?

File: src/main.cpp [hunk 22]
```diff
@@ -40,270 +26,264 @@ void setup(){
 
   LCD.begin(16,2) ;
 
-  Serial.begin(9600) ;
+  Serial.begin(115200) ;
```
Comment (wollip) [inline src/main.cpp]: debug

File: src/main.cpp [hunk 23]
```diff
@@ -40,270 +26,264 @@ void setup(){
...
   claw.fold(left_surface);
   delay(1000);
 
-  // claw.raise(left_surface ? sequences::RIGHT_CLAW : sequences::LEFT_CLAW);
-  // while(!claw.loop()) delay(100);
-  // claw.raise(left_surface ? sequences::LEFT_CLAW : sequences::RIGHT_CLAW);
-  // while(!claw.loop()) delay(100);
+  unsigned long waitUntil = millis() + loop_delay;
+  
+  enableEncoderInterrupts();
```
Comment (wollip) [inline src/main.cpp]: move line 58 to 60 to after `Insert extra variable here`. This way delays in the intialization won't affect things

File: src/main.cpp [hunk 24]
```diff
@@ -40,270 +26,264 @@ void setup(){
...
-    case 4:
-      driver.stop();
-  }
+// EXTRA VARIABLE
+  int state = 0;
+// CODE
+    if (stopbutton()) {
+      zipline.update();
+      platform.update();
+      state = 0;  // restart routine after update
```
Comment (wollip) [inline src/main.cpp]: reset `waitTime`

File: src/main.cpp [hunk 25]
```diff
@@ -40,270 +26,264 @@ void setup(){
...
-  }
-  delay(loop_delay);
-}
+// EXTRA VARIABLES
+  int state = 0;
+  unsigned int gate_time = 0;
+
+// CODE
+    if (stopbutton()){
+      gate.update();
```
Comment (wollip) [inline src/main.cpp]: reset `WaitTime`

File: src/main.cpp [hunk 26]
```diff
@@ -40,270 +26,264 @@ void setup(){
...
-        state++;
-      }
-      break;
-    case 3:  // lowering
-      if (platform.loop())
+// CODE
+    if (stopbutton())
+    {
+      platform.update();
+      state = 0;  // restart routine after updating
```
Comment (wollip) [inline src/main.cpp]: reset `WaitTime`

File: src/main.cpp [hunk 27]
```diff
@@ -40,270 +26,264 @@ void setup(){
...
+// EXTRA VARIABLE
+
+// CODE
+    pickup.loop();
+
+    if (stopbutton())
+    {
+      driver.stop();
+      pickup.update();
+      qrd.update();
```
Comment (wollip) [inline src/main.cpp]: reset `WaitTime`

File: src/main.cpp [hunk 28]
```diff
@@ -40,270 +26,264 @@ void setup(){
...
-  if (stopbutton()){
-    tape.update();
-    qrd.update();
-  }
-  delay(100);
-}
+    
+
+    if (stopbutton()){
+      qrd.update();
```
Comment (wollip) [inline src/main.cpp]: reset `WaitTime`

File: src/main.cpp [hunk 29]
```diff
@@ -40,270 +26,264 @@ void setup(){
...
+      case 3:  // finding zipline
+         if (zipline.loop())
+           state++;
+         break;
+    }
+  ///////////////////////////////
+  //// Do not touch /////////////
+  ///////////////////////////////
+    while (millis() < waitUntil) continue;
+    waitUntil += loop_delay;
```
Comment (wollip) [inline src/main.cpp]: safer to do `waitUntil = millis() + loop_delay` just in case we have some blocking code somewhere?

File: src/main.cpp [hunk 30]
```diff
@@ -328,81 +308,91 @@ void loop(){
...
+        else if (mode == 2) LCD.print("left");
+        
         LCD.setCursor(0,1); LCD.print(distance);
+        delay(50);
       }
-      delay(50);
+ 
+      if (mode == 0) driver.commandDriveStraight(distance);
+      else if (mode == 1) driver.commandTurnRight(distance);
+      else driver.commandTurnLeft(distance);
```
Comment (wollip) [inline src/main.cpp]: reset `WaitTime`

File: src/main.cpp [hunk 31]
```diff
@@ -328,81 +308,91 @@ void loop(){
...
+      if (stopbutton())
+        claw.update();
+      state = knob(6);
+      left = knob(7);
+      LCD.clear(); LCD.home();
+      LCD.setCursor(0,0); LCD.print("state");
+      LCD.setCursor(7,0); LCD.print(state);
+      LCD.setCursor(0,1); LCD.print("left");
+      LCD.setCursor(7,1); LCD.print(left);
+      delay(100);
```
Comment (wollip) [inline src/main.cpp]: reset `WaitTime`

File: src/sequences/drivetrain.cpp [hunk 32]
```diff
@@ -0,0 +1,128 @@
...
+  if (m_command != DRIVE_LINE_FOLLOW) m_lineFollowPower = 0;
+  m_command = DRIVE_LINE_FOLLOW;
+}
+
+void Drivetrain::commandBeaconFollow() {
+  m_command = DRIVE_BEACON_FOLLOW;
+}
+
+bool Drivetrain::readyForCommand() {
+  return m_command == IDLE;
```
Comment (wollip) [inline src/sequences/drivetrain.cpp:61]: i think line follow and beacon follow counts as ready for command states

File: src/sequences/drivetrain.cpp [hunk 33]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+void Drivetrain::setVelocity(int16_t left, int16_t right) {
+  m_leftVelocity = left;
+  m_rightVelocity = right;
+  m_command = DRIVE_VELOCITY;
+}
+
+void Drivetrain::tick() {
+  if (m_command == IDLE) {
```
Comment (wollip) [inline src/sequences/drivetrain.cpp:83]: maybe use a switch case to look nicer?

File: src/sequences/drivetrain.cpp [hunk 34]
```diff
@@ -0,0 +1,128 @@
...
+    stop();
+  }
+  else if (m_command == DRIVE_ENCODER) {
+    bool leftDone = m_leftMotor.setPosition(m_leftMotorTarget, m_speed);
+    bool rightDone = m_rightMotor.setPosition(m_rightMotorTarget, m_speed);
+    if (leftDone && rightDone) m_command = IDLE;
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: If one wheel finish a lot earlier than the other wheel, then the robot would turn. This is bad

File: src/sequences/drivetrain.cpp [hunk 35]
```diff
@@ -0,0 +1,128 @@
...
+  }
+  else if (m_command == DRIVE_LINE_FOLLOW) {
+    m_lineFollowPower += 5;
+    if (m_lineFollowPower > m_lineFollowMaxPower) m_lineFollowPower = m_lineFollowMaxPower;
+    int16_t error = m_lineSensor.getTapeError();
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: lets use `const` for these

File: src/sequences/drivetrain.cpp [hunk 36]
```diff
@@ -0,0 +1,128 @@
...
+    int16_t steer = error * m_lineFollowGain;
+    int16_t leftPower = m_lineFollowPower - (steer > 0 ? abs(steer) : 0);
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: you don't need `abs` on this line

File: src/sequences/drivetrain.cpp [hunk 37]
```diff
@@ -0,0 +1,128 @@
...
+    int16_t rightPower = m_lineFollowPower - (steer < 0 ? abs(steer) : 0);
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: change the `-` to a `+` and remove `abs`

File: src/sequences/drivetrain.cpp [hunk 38]
```diff
@@ -0,0 +1,128 @@
...
+  else if (m_command == DRIVE_VELOCITY) {
+    m_leftMotor.setVelocity(m_leftVelocity);
+    m_rightMotor.setVelocity(m_rightVelocity);
+  }
+  else if (m_command == DRIVE_POWER) {
+    m_leftMotor.setPower(m_leftPower);
+    m_rightMotor.setPower(m_rightPower);
+  }
+  else if (m_command == DRIVE_BEACON_FOLLOW) {
+    int32_t left10k = detect_10khz(L_BEACON_SENSOR());
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: const

File: src/sequences/drivetrain.cpp [hunk 39]
```diff
@@ -0,0 +1,128 @@
...
+    int32_t right10k = detect_10khz(R_BEACON_SENSOR());
+    int16_t error = 0;
+    if (left10k - right10k > BEACON_UNCERTAINTY()) error = BEACON_GAIN_P();
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: can we not implement the logic here again and just use `beacon`

File: src/sequences/drivetrain.cpp [hunk 40]
```diff
@@ -0,0 +1,128 @@
...
+    if (right10k - left10k > BEACON_UNCERTAINTY()) error = -BEACON_GAIN_P();
+    m_leftMotor.setPower(LINE_FOLLOW_0_SPEED() - error);
+    m_rightMotor.setPower(LINE_FOLLOW_0_SPEED() + error);
+  }
+}
+
+void Drivetrain::printTo(Print& p) {
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: debug

File: src/stages/gate.cpp [hunk 41]
```diff
@@ -12,24 +12,24 @@ bool Gate::gate_high_ = false;
...
   switch (state_){
     case 0:  // save encoder's start value as reference
-      encoder_start_ = encoder_.get(hardware::R_ENCODER_);
+      encoder_start_ = encoder_.getPosition();
       ++state_;
     case 1:  // going towards gate
-      if (encoder_.get(hardware::R_ENCODER_) - encoder_start_ < distance_){
-        follower_.loop();
+      if (encoder_.getPosition() - encoder_start_ < distance_){
+        driver_.commandLineFollow(true);
```
Comment (wollip) [inline src/stages/gate.cpp]: change `true` to `1`

File: src/stages/gate.cpp (hunk 18 above)
Comment (wollip) [inline src/stages/gate.cpp]: change `true` to `1`

File: src/utils/pid.cpp [hunk 42]
```diff
@@ -0,0 +1,30 @@
...
+  float derivative = error - prevError;
+  integral += error;
+  float output = kP * error + kI * integral * (1. - kDecay) + kD * derivative;
+  integral *= kDecay;
+  prevError = error;
+  prevOutput = output;
+  return output;
+}
+
+void PidController::printTo(Print& p) {
```
Comment (wollip) [inline src/utils/pid.cpp]: debug

File: src/utils/pid.cpp [hunk 43]
```diff
@@ -0,0 +1,30 @@
...
+  p.print(prevError);
+  p.print(' ');
+  p.print(integral);
+  p.print(' ');
+  //p.print(prevOutput);
+}
+
+}
```
Comment (wollip) [inline src/utils/pid.cpp]: comment namespace

File: src/sequences/drivetrain.cpp [hunk 44]
```diff
@@ -0,0 +1,128 @@
...
+  p.print(m_lineFollowPid.kP);
+  p.print(' ');
+  p.print(m_lineFollowPid.kI);
+}
+
+}
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: namespace

File: include/utils/pid.h [hunk 45]
```diff
@@ -0,0 +1,22 @@
...
+private:
+  float prevError;
+  float integral;
+  float prevOutput;
+};
+
+}
```
Comment (wollip) [inline include/utils/pid.h]: namespace

File: include/sequences/drivetrain.h [hunk 46]
```diff
@@ -0,0 +1,53 @@
...
+  int16_t m_rightPower;
+  int16_t m_leftVelocity;
+  int16_t m_rightVelocity;
+
+  enum Command { IDLE, DRIVE_VELOCITY, DRIVE_POWER, DRIVE_ENCODER, DRIVE_LINE_FOLLOW, DRIVE_BEACON_FOLLOW };
+  Command m_command;
+
+};
+
+}
```
Comment (wollip) [inline include/sequences/drivetrain.h]: namespace

File: include/sequences/drivetrain.h [hunk 47]
```diff
@@ -0,0 +1,53 @@
...
+
+  int16_t m_leftPower;
+  int16_t m_rightPower;
+  int16_t m_leftVelocity;
+  int16_t m_rightVelocity;
+
+  enum Command { IDLE, DRIVE_VELOCITY, DRIVE_POWER, DRIVE_ENCODER, DRIVE_LINE_FOLLOW, DRIVE_BEACON_FOLLOW };
+  Command m_command;
+
+};
```
Comment (wollip) [inline include/sequences/drivetrain.h]: class

File: include/hardware/encodermotor.h [hunk 48]
```diff
@@ -0,0 +1,39 @@
...
+  int16_t m_velocityTarget;
+  int32_t m_positionTarget;
+  int16_t m_maxPower;
+  int16_t m_maxVelocity;
+  utils::PidController m_velocityPid;
+  utils::PidController m_positionPid;
+  
+};
+
+}
```
Comment (wollip) [inline include/hardware/encodermotor.h]: namespace

File: include/hardware/encodermotor.h [hunk 49]
```diff
@@ -0,0 +1,39 @@
...
+  static const int16_t m_accelerationLimit = 20;
+  int16_t m_powerTarget;
+  int16_t m_velocityTarget;
+  int32_t m_positionTarget;
+  int16_t m_maxPower;
+  int16_t m_maxVelocity;
+  utils::PidController m_velocityPid;
+  utils::PidController m_positionPid;
+  
+};
```
Comment (wollip) [inline include/hardware/encodermotor.h]: class

File: include/hardware/encoder.h [hunk 50]
```diff
@@ -1,43 +1,30 @@
...
+private:
+  int32_t getPositionPrivate();
+  uint8_t m_encoderId;
+  bool m_reverse;
+  int32_t m_currentPosition;
+  int32_t m_previousPosition;
+  static const int16_t m_numVelocitySamples = 5;
+  int16_t m_velocitySamples[m_numVelocitySamples];
+  uint8_t m_currentVelocitySample;
+};
```
Comment (wollip) [inline include/hardware/encoder.h]: class

File: include/hardware/encoder.h [hunk 51]
```diff
@@ -1,43 +1,30 @@
...
+
+}
```
Comment (wollip) [inline include/hardware/encoder.h]: namespace

**Review by wollip — COMMENTED** [review]
Please change

File: src/sequences/drivetrain.cpp [hunk 52]
```diff
@@ -0,0 +1,128 @@
...
+  if (m_command == IDLE) {
+    stop();
+  }
+  else if (m_command == DRIVE_ENCODER) {
+    bool leftDone = m_leftMotor.setPosition(m_leftMotorTarget, m_speed);
+    bool rightDone = m_rightMotor.setPosition(m_rightMotorTarget, m_speed);
+    if (leftDone && rightDone) m_command = IDLE;
+  }
+  else if (m_command == DRIVE_LINE_FOLLOW) {
+    m_lineFollowPower += 5;
```
Comment (wollip) [inline src/sequences/drivetrain.cpp]: Is this the "Acceleration limit"??!?!?!?!?!?!?

**Comment by wollip** [comment]
I will not agree with this change until i can do this without wheelies
```
while (true) {
  drivetrain.setPower(255,255) 
  delay()
  tick()
  drivetrain.setPower(-255,255)
  delay()
  tick()
}
```

**Review by wollip — COMMENTED** [review]
Please fix changes

File: include/configs.h (hunk 1 above)
Comment (wollip) [inline include/configs.h]: they are still being used by drivetrain.cpp

File: include/hardware/encodermotor.h [hunk 53]
```diff
@@ -0,0 +1,39 @@
...
+#include "hardware/encoder.h"
+#include "utils/pid.h"
+
+namespace hardware {
+
+class EncoderMotor {
+public:
+  EncoderMotor(int motor_num, Encoder& encoder, bool backward=false);
+  void setPower(int16_t power);
+  void setVelocity(int16_t velocity, int16_t power = 255);
```
Comment (wollip) [inline include/hardware/encodermotor.h]: this is fine

File: src/hardware/encodermotor.cpp (hunk 20 above)
Comment (wollip) [inline src/hardware/encodermotor.cpp]: this is fine

File: include/configs.h (hunk 1 above)
Comment (wollip) [reply include/configs.h:36]: they are still being used by drivetrain

**Review by fb39ca4 — APPROVED** [review]

**Pull request merged**